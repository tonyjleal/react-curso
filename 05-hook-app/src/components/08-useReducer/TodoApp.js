import React, { useEffect, useReducer } from 'react';
import { useForm } from '../../hooks/useForm';
import './style.css';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';


const init = () => {
   return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {

 const [ todos, dispatch ]  = useReducer(todoReducer, [], init);
 


  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleAddTodo = ( newTodo ) => {
    dispatch({
      type: 'add',
      payload: newTodo
    });
  }

  const handleDelete = ( todoId ) => {

    const action = {
      type: 'delete',
      payload: todoId
    }

    dispatch(action);
  }

  const handleToggle = (todoId) => {

    const action = {
      type: 'toggle', 
      payload: todoId
    }

    dispatch(action);
  }

  return(
     <div>

        <h1> TODO APP ( { todos.length } )</h1>
        <hr />
        <div  className='row'
        >
          <div className='col-7'>
              <TodoList 
                todos={todos} 
                handleDelete={handleDelete} 
                handleToggle={handleToggle}/>
          </div>
          <div className='col-5'>
              <TodoAdd 
                handleAddTodo={handleAddTodo}/>
          </div>
        </div>
    </div>
    );
};
