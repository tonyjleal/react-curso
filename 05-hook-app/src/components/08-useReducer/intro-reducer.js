
const initialState = [{
    id: 1,
    todo: 'Buy milk',
    done: false,
}]


const todoReducer = ( state = initialState, action ) => {

    if (action?.type === 'add') {
        return [...state, action.payload];
    }

    return state;
}

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Buy cookies',
    done: false,
}

const action = {
    type: 'add',
    payload: newTodo
}

todos = todoReducer(todos, action)

console.log(todos);