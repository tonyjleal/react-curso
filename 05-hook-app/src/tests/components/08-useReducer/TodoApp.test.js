import { shallow, mount } from 'enzyme'
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../fixtures/demoTodos';
import { act } from '@testing-library/react'

describe('Test <TodoApp />', () => { 
    
    const wrapper = shallow(<TodoApp />);
    beforeEach( () => {
        Storage.prototype.setItem = jest.fn(() => {});
    })
 

    test('should show component success', () => { 
        
        expect(wrapper).toMatchSnapshot();

     })
    
     test('should add TODO', () => { 
        
        const wrapper = mount(<TodoApp />);

        act( () => {
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0])
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1])
        })

        const h1 = wrapper.find('h1');
        expect(h1.text().trim()).toBe('TODO APP ( 2 )');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);

     });

     test('should delete a TODO', () => { 
         
        act(() => {
 
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
            wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);
 
        })
        expect(wrapper.find('h1').text().trim()).toBe('TODO APP ( 0 )');
 
      })
 })

