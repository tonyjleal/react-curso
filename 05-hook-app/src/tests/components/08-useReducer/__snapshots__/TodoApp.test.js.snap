// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <TodoApp /> should show component success 1`] = `
<div>
  <h1>
     TODO APP ( 
    0
     )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
